### ISSUE TITLE
Reject large size block via http rpc

### ISSUE THREAD

**Issue by sifnoc** [zkopru-network/zkopru#343]
When the coodirnator propose a large size block(over 200 zktx included, each zktx 1 inflow 2 outflow), go-ethereum node return error with "oversized data".

geth http rpc has a size limit of the transaction

https://github.com/ethereum/go-ethereum/blob/fc01a7ce8e4f865b0d282c21a42c21b0bd1d0b0c/rpc/http.go#L35

possible to avoid this with using ipc providor

**Comment by JChanceHud** [comment]
Good catch! This affects the economics of running a coordinator. I guess for now we just limit the total size in the `block-generator`.

**Comment by sifnoc** [comment]
we already implemented the transaction size limit on the `tx-pool`, but it looks not accurate.

most coordinators use default `maxBytes` as 131072(= 1024 * 128), it is the same number a single transaction bytes limit on `tx-pool` of geth.

https://github.com/zkopru-network/zkopru/blob/03a6348f80dd5128855ce17c555b50bffcb13715/packages/coordinator/src/tx-pool.ts#L259-L262

In here, `available` in the remaining bytes after adding tx from `tx-pool` in zkopru.

I measured the size of a single zkTx which has one inflow two outflows, it was 589. so, the biggest layer2 block that can be generated by the coordinator is 222 ZkTxs included(if one inflow, one outflow can be more included)

As follow code, It seems that the geth tx-pool limits a transaction by whole transaction bytes, not an input data size.

https://github.com/ethereum/go-ethereum/blob/9489853321bb221694f5262772c656413be073af/core/types/transaction.go#L379-L387
```go
type Transaction struct {
	inner TxData    // Consensus contents of a transaction
	time  time.Time // Time first seen locally (spam avoidance)

	// caches
	hash atomic.Value
	size atomic.Value
	from atomic.Value
}

type TxData interface {
	txType() byte // returns the type ID
	copy() TxData // creates a deep copy and initializes all fields

	chainID() *big.Int
	accessList() AccessList
	data() []byte
	gas() uint64
	gasPrice() *big.Int
	gasTipCap() *big.Int
	gasFeeCap() *big.Int
	value() *big.Int
	nonce() uint64
	to() *common.Address

	rawSignatureValues() (v, r, s *big.Int)
	setSignatureValues(chainID, v, r, s *big.Int)
}

...

func (tx *Transaction) Size() common.StorageSize {
	if size := tx.size.Load(); size != nil {
		return size.(common.StorageSize)
	}
	c := writeCounter(0)
	rlp.Encode(&c, &tx.inner)
	tx.size.Store(common.StorageSize(c))
	return common.StorageSize(c)
}
```

so, If it is right, we should reduce `maxBytes` as much as signature and other fields.

**Comment by sifnoc** [comment]
Checked actual size of a propose tx from the coordinator

the propose tx has 1 zktx(one inflow, two outflows): 1,233 bytes
108 bytes for `txParams` excluding data(`from`, `to`, `value`, `signatures` and others)
1,125 bytes for data(propose header, the `zkTx`, `parentBlock.hash` and other)

the `zkTx`(one inflow, two outflows) is only 591 bytes.
And remains data is,
4 bytes for `safePropose` function signature,
340 bytes for serialized header of layer 2 block,
32 bytes for parent block hash,
158 bytes for unknown

the unknown data size varies, 132 ~ 159 bytes, it changed on how many zktx are included. but not linearly increasing or decreasing.

For now, we can avoid this issue by increasing `consumedBytes`in the `block-generator`.
https://github.com/zkopru-network/zkopru/blob/03a6348f80dd5128855ce17c555b50bffcb13715/packages/coordinator/src/middlewares/default/block-generator.ts#L25-L26

Roughly, `760` bytes will be enough.
